data_structures/Heap/Binary_Heap/min_heap_two.dart

```diff
@@ -0,0 +1,113 @@
+import 'package:test/test.dart';
+
+class MinHeap {
+  List<int> heap;
+
+  void buildHeap(List<int> array) {
+    this.heap = _heapify(array);
+  }
+
+  List<int> _heapify(List<int> array) {
+    int firstParent;
+    firstParent = (array.length - 2) ~/ 2;
+    for (int i = firstParent; i >= 0; i--) {
+      _siftDown(i, array.length - 1, array);
+    }
+    return array;
+  }
+
+  int peek() {
+    if (!isEmpty()) {
+      return this.heap[0];
+    }
+    return null;
+  }
+
+  bool isEmpty() {
+    return this.heap.length == 0;
+  }
+
+  void _siftUp(int currentIndex) {
+    int parentIndex;
+    parentIndex = (currentIndex - 1) ~/ 2;
+    while (
+        parentIndex >= 0 && this.heap[parentIndex] > this.heap[currentIndex]) {
+      _swap(parentIndex, currentIndex, this.heap);
+      currentIndex = parentIndex;
+      parentIndex = (currentIndex - 1) ~/ 2;
+    }
+  }
+
+  void _siftDown(int currentIndex, int endIndex, List<int> heap) {
+    int childOneIndex = (2 * currentIndex) + 1;
+    int childTwoIndex;
+
+    while (childOneIndex <= endIndex) {
+      childTwoIndex =
+          2 * currentIndex + 2 <= endIndex ? 2 * currentIndex + 2 : -1;
+      int indexToSwap;
+      if (childTwoIndex != -1 && heap[childTwoIndex] < heap[childOneIndex]) {
+        indexToSwap = childTwoIndex;
+      } else {
+        indexToSwap = childOneIndex;
+      }
+
+      if (heap[currentIndex] > heap[indexToSwap]) {
+        _swap(currentIndex, indexToSwap, heap);
+        currentIndex = indexToSwap;
+        childOneIndex = (2 * currentIndex) + 1;
+      } else {
+        break;
+      }
+    }
+  }
+
+  void insert(int value) {
+    this.heap.add(value);
+    _siftUp(this.heap.length - 1);
+  }
+
+  int remove() {
+    if (!isEmpty()) {
+      _swap(0, this.heap.length - 1, this.heap);
+      int minElement = this.heap.removeLast();
+      _siftDown(0, this.heap.length - 1, this.heap);
+      return minElement;
+    }
+    return null;
+  }
+
+  void _swap(int left, int right, List<int> array) {
+    int temp;
+    temp = array[left];
+    array[left] = array[right];
+    array[right] = temp;
+  }
+}
+
+void main() {
+  MinHeap minheap = new MinHeap();
+  List<int> array = [48, 12, 24, 7, 8, -5, 24, 391, 24, 56, 2, 6, 8, 41];
+  minheap.buildHeap(array);
```

Comment:
Could you create those variables inside the test cases?
I've got confused when i seen 2 in the last test case - unit tests should not depend on side effects of previous ones.

Reply:
I was thinking about updating existing test cases rather than creating new one.
For example
```dart
  test(('insert and remove'), () {
    List<int>array = [-7, 2, 3, 8, -10, 4, -6, -10, -2, -7, 10, 5, 2, 9, -9, -5, 3, 8];
    MinHeap minheap = new MinHeap();
    minheap.buildHeap(array);
    minheap.insert(-100);
    minheap.insert(-100);

    expect(minheap.remove(), equals(-100));
    expect(minheap.remove(), equals(-100));
  });
```
When `minheap` is shared between all test cases this can lead for ugly bugs - for example if you change test case order. Or if you try to run tests in parallel.